cfme/tests/intelligence/test_chargeback.py

```diff
@@ -26,85 +25,101 @@ def with_random_per_time(**kw):
             fields={
                 'Allocated CPU Count': with_random_per_time(fixed_rate='1000'),
                 'Used Disk I/O': with_random_per_time(fixed_rate='10'),
                 'Fixed Compute Cost 1': with_random_per_time(fixed_rate='100'),
-                'Used Memory': with_random_per_time(**rate[0]),
-                'Used CPU Cores': with_random_per_time(**rate[1]),
-            })
+                'Used Memory': with_random_per_time(**rate_values[0]),
+                'Used CPU Cores': with_random_per_time(**rate_values[1]),
+            }
+        )
 
     elif rate_resource == 'storage':
-        rate = cb.StorageRate(
-            description='cb_{}_{}_{}_{}'.format(fauxfactory.gen_alphanumeric(), rate_type,
-                rate_resource, rate_action),
+        rate = appliance.collections.storage_rates.create(
+            description=rate_description,
             fields={
                 'Fixed Storage Cost 1': with_random_per_time(fixed_rate='100'),
                 'Fixed Storage Cost 2': with_random_per_time(fixed_rate='300'),
-                'Allocated Disk Storage': with_random_per_time(**rate[0]),
-                'Used Disk Storage': with_random_per_time(**rate[1]),
-            })
+                'Allocated Disk Storage': with_random_per_time(**rate_values[0]),
+                'Used Disk Storage': with_random_per_time(**rate_values[1]),
+            }
+        )
+
+    yield rate
 
-    @request.addfinalizer
-    def _cleanup():
-        if rate.exists:
-            rate.delete()
-    return rate
+    if rate.exists:
+        rate.delete()
 
 
-def test_compute_chargeback_duplicate_disallowed(request):
+@pytest.mark.parametrize('rate_resource', ['compute', 'storage'])
+@pytest.mark.parametrize('rate_type', ['fixed', 'variable'])
+@pytest.mark.parametrize('rate_action', ['add'])
+def test_chargeback_duplicate_disallowed(
+    request, chargeback_rate, rate_resource, rate_type, rate_action, appliance
+):
     """
     Polarion:
         assignee: nachandr
         casecomponent: CandU
         caseimportance: low
         initialEstimate: 1/12h
     """
-    cb_rate = chargeback_rate('compute', 'fixed', 'add', request)
-
-    cb_rate.create()
+    cb_rate = chargeback_rate  # for brevity
     assert cb_rate.exists
     with pytest.raises(AssertionError):
-        cb_rate.create()
+        if rate_resource == 'compute':
+            appliance.collections.compute_rates.create(
+                description=cb_rate.description,
+                fields=cb_rate.fields
+            )
+        elif rate_resource == 'storage':
+            appliance.collections.storage_rates.create(
+                description=cb_rate.description,
+                fields=cb_rate.fields
+            )
     # view should still be on the add form
-    view = cb_rate.create_view(
-        navigator.get_class(cb_rate, 'Add').VIEW)
+    view = cb_rate.create_view(navigator.get_class(cb_rate.parent, 'Add').VIEW)
```

Comment:
same comment @izapolsk had about using `wait` here

Reply:
I did this in a few other places in this module now too.